README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# simK <img src="man/figures/logo.png" height="150" align="right"/>

<!-- badges: start -->

[![](https://img.shields.io/badge/devel%20version-0.2.1-blue.svg)](https://github.com/txopen/simK)
[![R-CMD-check](https://github.com/txopen/simK/workflows/R-CMD-check/badge.svg)](https://github.com/txopen/simK/actions)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![License:
MIT](https://img.shields.io/badge/license-MIT-blue.svg)](https://cran.r-project.org/web/licenses/MIT)
[![Codecov test
coverage](https://codecov.io/gh/txopen/simK/branch/main/graph/badge.svg)](https://app.codecov.io/gh/txopen/simK?branch=main)
<!-- badges: end -->

:dart: Functions to procedurally generate synthetic data in R for kidney
transplant simulations. You can see `simK` in action,
[here](https://txopen.github.io/simK/index.html).

`simK` allows to generate data with clinical and demographic information
for a pool of simulated cadaveric donors and simulated wait listed
candidates for kidney transplantation.

Synthetic data can be used for preliminary analysis and the development
of analysis code.

Data generated with `simK` are particularly useful on
[KARS](https://balima.shinyapps.io/kars/).

> :warning: This package is not a medical device and <ins>should not be
> used for making clinical decisions</ins>.

## Instalation

### Development version

The development version can be installed from GitHub, if you want all
the latest features, together with all the latest bugs and errors. You
have been warned :exclamation:

    # install from GitHub
    devtools::install_github("txopen/simK")

## Main functions

This package has 3 main functions, with them we can generate simulated
data for a pool of donors, a set of kidney transplant candidates and the
respective HLA-antibodies for those patients HLA sensitized.

A detailed description of these functions usage, is available from
[here](https://txopen.github.io/simK/articles/simk.html).

:warning: For PT origin, all these functions rely on HLA typing at
intermediate resolution as described at [Lima *et al*,
2013](https://12f11c1f-960a-f627-594d-b8ce276384f7.filesusr.com/ugd/3e838e_dc548dede99a4db5869c3d2c20c2d16f.pdf?index=true).

:warning: For NMDP populations, HLA typing were described by [Gragert
*et al*,
2013](https://reader.elsevier.com/reader/sd/pii/S0198885913001821?token=885F05F0FAD857548C932A5CC40B6962CD7A9FC589A73FC6DECE56D6E4E247587E06D755C16D9F2EA425111289B5DC2F&originRegion=eu-west-1&originCreation=20220222173605)


NEWS.md
# simK 0.2.1

* corrected functions arguments' names

* increased unit-test quality and coverage.

# simK 0.2.0

* now depends on R > 4.1 and uses the native pipe

# simK 0.1.3

* added input `seed.number` to ensure reproducibility

# simk 0.1.2

* added `origin` options that allows simulate also HLA tipyng from MNDP 

# simK 0.1.1

* changed functions names to snake case

* added require(magrittr) to functions using %>% 

# simK 0.1.0

* first release version


tests/testthat.R
library(testthat)
library(simK)

test_check("simK")


tests/testthat/test_dfs.R
# test_that("data freame with donors", {
#   result1 <-  tibble::tribble(
#     ~ID, ~bg, ~A1, ~A2, ~B1, ~B2, ~DR1, ~DR2, ~age,
#     'D1','A','1','29','8','44','11','7',53,
#     'D2','O','2','3','7','57','4','8',57,
#     'D3','A','11','33','14','35','1','13',61,
#     'D4','O','24','30','49','58','11','11',62,
#     'D5','B','2','3','7','51','1','13',66,
#     )
#   result2 <-  tibble::tribble(
#     ~ID, ~bg, ~A1, ~A2, ~B1, ~B2, ~DR1, ~DR2, ~age,
#     'D1','','','','','','','','',
#     'D2','','','','','','','','',
#     'D3','','','','','','','','',
#     'D4','','','','','','','','',
#     'D5','','','','','','','','',
#   )
#   result3 <-  tibble::tribble(
#     ~ID, ~bg, ~A1, ~A2, ~B1, ~B2, ~DR1, ~DR2, ~age,
#     'D1','','','','','','','','',
#     'D2','','','','','','','','',
#     'D3','','','','','','','','',
#     'D4','','','','','','','','',
#     'D5','','','','','','','','',
#   )
#   result4 <-  tibble::tribble(
#     ~ID, ~bg, ~A1, ~A2, ~B1, ~B2, ~DR1, ~DR2, ~age,
#     'D1','','','','','','','','',
#     'D2','','','','','','','','',
#     'D3','','','','','','','','',
#     'D4','','','','','','','','',
#     'D5','','','','','','','','',
#   )|>as.data.frame()
#
#
#
#   res0 <- donors_df(n=5)
#   expect_identical(res0,
#                    result1)
#
# })


candidates_df(n=100) %>%
  dplyr::count(urgent)

hla_sample(n = 1000, replace = TRUE, origin = 'API')

hla_sample_mndp(n = 1000, replace = TRUE, origin = 'CAU')


tests/testthat/test_histo.R
test_that("number of HLA mm", {
  expect_equal(mmHLA(dA = c('1','2'), dB = c('5','7'), dDR = c('1','4'),
                     cA = c('1','2'), cB = c('3','15'), cDR = c('4','7')),
               c(mmA = 0, mmB = 2, mmDR = 1, mmHLA = 3))
  expect_equal(mmHLA(dA = c('1','2'), dB = c('5','7'), dDR = c('4','4'),
                     cA = c('3','4'), cB = c('5','15'), cDR = c('4','7')),
               c(mmA = 2, mmB = 1, mmDR = 0, mmHLA = 3))
  expect_equal(mmHLA()["mmHLA"][[1]],
               (mmHLA()["mmA"] + mmHLA()["mmB"] + mmHLA()["mmDR"])[[1]])

  expect_error(mmHLA(dA = c(1,2)))
  expect_error(mmHLA(dB = c(1,2)))
  expect_error(mmHLA(dDR = c(1,2)))
  expect_error(mmHLA(cA = c(1,2)))
  expect_error(mmHLA(cB = c(1,2)))
  expect_error(mmHLA(dcDR = c(1,2)))
})

test_that("Matchability from D10K", {
  expect_equal(matchability(cABO = 'A', cPRA = 85, seed.number = 3),
               67)
  expect_equal(matchability(cABO = 'B', cPRA = 85, seed.number = 3),
               10)
  expect_equal(matchability(cABO = 'AB', cPRA = 50, seed.number = 1),
               13)
  expect_equal(matchability(cABO = 'O', cPRA = 0, seed.number = 1),
               395)

  invalid.ABO <- list(0, 'X', 'a', 'o', 'b')
  for(i in 1:length(invalid.ABO)){
  expect_error(matchability(cABO = invalid.ABO[[i]]))
  }
})

test_that("virtual PRA", {
  expect_equal(vpra(abs = c('A1','A2','B5','DR4'), donors = D10K),
               70)
  expect_equal(vpra(abs = c('A11','A30','DR8'), donors = D10K),
               24.7)
  expect_equal(vpra(abs = c('DR13'), donors = D10K),
               28.5)
  expect_equal(vpra(abs = c('A29'), donors = D10K),
               10.9)
})

test_that("HLA antibodies", {
  expect_equal(antbs(cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
                     cPRA = 85,
                     origin = 'PT', seed.number = 3)$vPRA,
               82.8)
  expect_equal(antbs(cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
                     cPRA = 50,
                     origin = 'PT', seed.number = 1)$vPRA,
               50.1)
  expect_equal(antbs(cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
                     cPRA = 50,
                     origin = 'PT', seed.number = 1)$Abs,
               c("DR15","A24","B53","A1"))

  invalid.origin <- list('p',0,1,'us','api','afa','c')
  for(i in 1:length(invalid.origin)){
    expect_error(antbs(origin = invalid.origin[[i]]))
    }
})


R/data.R
#' HLA typing.
#'
#' A dataset containing HLA typing from 37993 unrelated voluntary bone marrow donors from Nothern Portugal, recruited between 2010 and 2011.
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{C1}{HLA-C allele 1}
#'   \item{C2}{HLA-C allele 2}
#'   \item{DRB11}{HLA-DRB1 allele 1}
#'   \item{DRB12}{HLA-DRB1 allele 2}
#'   ...
#' }
#' @source \url{https://12f11c1f-960a-f627-594d-b8ce276384f7.filesusr.com/ugd/3e838e_dc548dede99a4db5869c3d2c20c2d16f.pdf?index=true}
#' @concept data
"hla"


#' A pool of 10.0000 donors.
#'
#' A synthetic dataset with information for 10.000 donors, generated with `donors.df()` and `set.seed(3)` .
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{bg}{ABO blood group}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{DR1}{HLA-DRB1 allele 1}
#'   \item{DR2}{HLA-DRB1 allele 2}
#'   \item{age}{age in years}
#'   ...
#' }
#' @source \url{https://12f11c1f-960a-f627-594d-b8ce276384f7.filesusr.com/ugd/3e838e_dc548dede99a4db5869c3d2c20c2d16f.pdf?index=true}
#' @concept data
"D10K"


#' A vector with HLA-A.
#'
#' A character vector with antigens for HLA-A locus obtained from \code{hla} dataset.
#'
#' @format A character vector with length 20.
#' @concept data
"agA"

#' A vector with HLA-B.
#'
#' A character vector with antigens for HLA-B locus obtained from \code{hla} dataset.
#'
#' @format A character vector with length 34.
#' @concept data
"agB"

#' A vector with HLA-DR.
#'
#' A character vector with antigens for HLA-DR locus obtained from \code{hla} dataset.
#'
#' @format A character vector with length 13.
#' @concept data
"agDR"


#' NMDP HLA frequencies.
#'
#' A dataset containing Frequencies of HLA-A, -B, -DR haplotypes estimated (Expectation Maximization) by race/ethnicity from a subset of donors in the NMDP registry.
#'
#' @format A data frame with 8696 rows and 7 variables:
#' \describe{
#'   \item{A}{HLA-A allele}
#'   \item{B}{HLA-B allele}
#'   \item{DR}{HLA-DRB1 allele}
#'   \item{api}{haplotye frequencies for Asian / Pacific Islander}
#'   \item{afa}{haplotye frequencies for African American / Black}
#'   \item{cau}{haplotye frequencies for Withe Caucasian}
#'   \item{his}{haplotye frequencies for Hispanic}
#'   ...
#' }
#' @source \url{https://bioinformatics.bethematchclinical.org/hla-resources/haplotype-frequencies/a-b-drb1-224-haplotype-frequencies/}
#' @concept data
"MNDPhaps"

#' A vector with HLA-A from MNDP.
#'
#' A character vector with  HLA-A locus antigens, obtained from \code{MNDPhaps} dataset.
#'
#' @format A character vector with length 21.
#' @concept data
"agA_MNDP"

#' A vector with HLA-B from MNDP.
#'
#' A character vector with  HLA-B locus antigens, obtained from \code{MNDPhaps} dataset.
#'
#' @format A character vector with length 41.
#' @concept data
"agB_MNDP"

#' A vector with HLA-DR from MNDP.
#'
#' A character vector with  HLA-DR locus antigens, obtained from \code{MNDPhaps} dataset.
#'
#' @format A character vector with length 13.
#' @concept data
"agDR_MNDP"

#' A pool of 10.0000 donors API.
#'
#' A synthetic dataset with information for 10.000 donors, generated with \code{donors.df()} and \code{set.seed(3)} from MNDP's Asian / Pacific Islander population.
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{bg}{ABO blood group}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{DR1}{HLA-DRB1 allele 1}
#'   \item{DR2}{HLA-DRB1 allele 2}
#'   \item{age}{age in years}
#'   ...
#' }
#' @source \url{https://bioinformatics.bethematchclinical.org/hla-resources/haplotype-frequencies/a-b-drb1-224-haplotype-frequencies/}
#' @concept data
"D10K_API"

#' A pool of 10.0000 donors AFA.
#'
#' A synthetic dataset with information for 10.000 donors, generated with \code{donors.df()} and \code{set.seed(3)} from MNDP's African American / Black population.
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{bg}{ABO blood group}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{DR1}{HLA-DRB1 allele 1}
#'   \item{DR2}{HLA-DRB1 allele 2}
#'   \item{age}{age in years}
#'   ...
#' }
#' @source \url{https://bioinformatics.bethematchclinical.org/hla-resources/haplotype-frequencies/a-b-drb1-224-haplotype-frequencies/}
#' @concept data
"D10K_AFA"

#' A pool of 10.0000 donors CAU.
#'
#' A synthetic dataset with information for 10.000 donors, generated with \code{donors.df()} and \code{set.seed(3)} from MNDP's White / Caucasian population.
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{bg}{ABO blood group}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{DR1}{HLA-DRB1 allele 1}
#'   \item{DR2}{HLA-DRB1 allele 2}
#'   \item{age}{age in years}
#'   ...
#' }
#' @source \url{https://bioinformatics.bethematchclinical.org/hla-resources/haplotype-frequencies/a-b-drb1-224-haplotype-frequencies/}
#' @concept data
"D10K_CAU"

#' A pool of 10.0000 donors HIS.
#'
#' A synthetic dataset with information for 10.000 donors, generated with \code{donors.df()} and \code{set.seed(3)} from MNDP's Hispanic population.
#'
#' @format A data frame with 37993 rows and 9 variables:
#' \describe{
#'   \item{ID}{Identifier}
#'   \item{bg}{ABO blood group}
#'   \item{A1}{HLA-A allele 1}
#'   \item{A2}{HLA-A allele 2}
#'   \item{B1}{HLA-B allele 1}
#'   \item{B2}{HLA-B allele 2}
#'   \item{DR1}{HLA-DRB1 allele 1}
#'   \item{DR2}{HLA-DRB1 allele 2}
#'   \item{age}{age in years}
#'   ...
#' }
#' @source \url{https://bioinformatics.bethematchclinical.org/hla-resources/haplotype-frequencies/a-b-drb1-224-haplotype-frequencies/}
#' @concept data
"D10K_HIS"


R/demographic.R
#' Builds a vector with truncated ages
#'
#' @description Returns a vector with ages with a normal distribution truncated
#' by upper and upper limits.
#' @param n An integer to define the length of the returned vector.
#' @param lower An integer for ages' lower limit.
#' @param upper An integer for ages' upper limit.
#' @param mean A value for mean's distribution.
#' @param sd A value for standar deviation's distribution.
#' @param seed.number Seed for pseudo random number generator.
#' @return A vector length `n` with normal distributed truncated ages.
#' @examples
#' ages(n = 100, lower=18, upper=75, mean = 55, sd = 15, seed.number = 123)
#' @export
#' @concept demographic_parameters
ages <- function(n = 100, lower=18, upper=75, mean = 55, sd = 15, seed.number = 123){

  if(!is.numeric(n) | n < 1){stop("`n` must be a positive single number!\n")}
  if(!is.numeric(mean) | mean < 1){stop("`mean` must be a positive single number!\n")}
  if(!is.numeric(sd) | sd < 1){stop("`sd` must be a positive single number!\n")}
  if(!is.numeric(lower) | lower < 1){stop("`lower` must be a positive single number!\n")}
  if(!is.numeric(upper) | upper < 1){stop("`upper` must be a positive single number!\n")}

  set.seed(seed.number)

  round(
    truncnorm::rtruncnorm(n = n, a=lower, b=upper, mean = mean, sd = sd)
  )

}





R/histo.R
#' number of HLA mismatchs
#'
#' @description Computes the number of HLA mismatchs between one donor and one candidate
#' @param dA donor's HLA-A typing
#' @param dB donor's HLA-B typing
#' @param dDR donor's HLA-DR typing
#' @param cA candidate's HLA-A typing
#' @param cB candidate's HLA-B typing
#' @param cDR candidate's HLA-DR typing
#' @return mmA number of HLA-A mismatchs between \code{dA} and \code{cA};
#' mmB number of HLA-B mismatchs between \code{dB} and \code{cB};
#' mmDR number of HLA-DR mismatchs between \code{dDR} and \code{cDR};
#' and mmHLA as the sum of mmA + mmB + mmDR
#' @examples
#' mmHLA(dA = c('1','2'), dB = c('5','7'), dDR = c('1','4'),
#' cA = c('1','2'), cB = c('03','15'), cDR = c('04','07'))
#' @export
#' @concept histocompatibility
mmHLA <- function(dA = c('1','2'), dB = c('5','7'), dDR = c('1','4'),
                  cA = c('1','2'), cB = c('3','15'), cDR = c('4','7')){

    if(!is.character(dA)){stop("donor's HLA-A typing is not valid!\n")}
    if(!is.character(dB)){stop("donor's HLA-B typing is not valid!\n")}
    if(!is.character(dDR)){stop("donor's HLA-DR typing is not valid!\n")}
    if(!is.character(cA)){stop("candidate's HLA-A typing is not valid!\n")}
    if(!is.character(cB)){stop("candidate's HLA-B typing is not valid!\n")}
    if(!is.character(cDR)){stop("candidate's HLA-DR typing is not valid!\n")}

    mmA <- NULL
    mmB <- NULL
    mmDR <- NULL

    # compute missmatches
    mmA <- dplyr::if_else((dA[1] %in% cA & dA[2] %in% cA) | (dA[1] %in% cA & (is.na(dA[2]) | dA[2] == "")), 0,
                          dplyr::if_else(dA[1] %in% cA | dA[2] %in% cA, 1,
                                         dplyr::if_else(!dA[1] %in% cA & (is.na(dA[2]) | dA[2] == ""), 1,
                                                        dplyr::if_else(dA[1] == dA[2], 1,2))))

    mmB <- dplyr::if_else((dB[1] %in% cB & dB[2] %in% cB) | (dB[1] %in% cB & (is.na(dB[2]) | dB[2] == "")), 0,
                          dplyr::if_else(dB[1] %in% cB | dB[2] %in% cB, 1,
                                         dplyr::if_else(!dB[1] %in% cB & (is.na(dB[2]) | dB[2] == ""), 1,
                                                        dplyr::if_else(dB[1] == dB[2], 1,2))))

    mmDR <- dplyr::if_else((dDR[1] %in% cDR & dDR[2] %in% cDR) | (dDR[1] %in% cDR & (is.na(dDR[2]) | dDR[2] == "")), 0,
                           dplyr::if_else(dDR[1] %in% cDR | dDR[2] %in% cDR, 1,
                                          dplyr::if_else(!dDR[1] %in% cDR & (is.na(dDR[2]) | dDR[2] == ""), 1,
                                                         dplyr::if_else(dDR[1] == dDR[2],1,2))))

    mmHLA <- mmA + mmB + mmDR
    mm <- c(mmA,mmB,mmDR,mmHLA)
    names(mm) <- c("mmA","mmB","mmDR","mmHLA")

    return(mm)


  # # compute missmatches with base ifelse
  # mmA<-ifelse((dA[1] %in% cA & dA[2] %in% cA) | (dA[1] %in% cA & (is.na(dA[2]) | dA[2] == "")), 0,
  #             ifelse(dA[1] %in% cA | dA[2] %in% cA, 1,
  #                    ifelse(!dA[1] %in% cA & (is.na(dA[2]) | dA[2] == ""), 1,
  #                           ifelse(dA[1] == dA[2], 1,2))))
  #
  # mmB<-ifelse((dB[1] %in% cB & dB[2] %in% cB) | (dB[1] %in% cB & (is.na(dB[2]) | dB[2] == "")), 0,
  #             ifelse(dB[1] %in% cB | dB[2] %in% cB, 1,
  #                    ifelse(!dB[1] %in% cB & (is.na(dB[2]) | dB[2] == ""), 1,
  #                           ifelse(dB[1] == dB[2], 1,2))))
  #
  # mmDR<-ifelse((dDR[1] %in% cDR & dDR[2] %in% cDR) | (dDR[1] %in% cDR & (is.na(dDR[2]) | dDR[2] == "")), 0,
  #              ifelse(dDR[1] %in% cDR | dDR[2] %in% cDR, 1,
  #                     ifelse(!dDR[1] %in% cDR & (is.na(dDR[2]) | dDR[2] == ""), 1,
  #                            ifelse(dDR[1] == dDR[2],1,2))))

}

#' Matchability from D10K
#'
#' @description Computes the number donors on dataset D10K that are a match to
#' a given transplant candidate. A sample of D10K is selected according to
#' cPRA value, afterwards donors ABO identical and HLA mismatch level 1 or 2
#' (0 DR or (1 DR and 0 B)) are filtered.
#' @param cABO A character from 'A', 'B', 'AB', 'O'
#' @param cPRA candidate's cPRA value
#' @param cA candidate's HLA-A typing
#' @param cB candidate's HLA-B typing
#' @param cDR candidate's HLA-DR typing
#' @param seed.number a numeric seed that will be used for random number generation.
#' @return Match Score measure of how difficult it is to match a patient with a organ donor. A score from 1 (easy to match) to 10 (difficult to match).
#' @examples
#' matchability(cABO = 'A', cPRA = 85,
#' cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
#' seed.number = 3)
#' @export
#' @concept histocompatibility
matchability <- function(cABO = 'A', cPRA = 85,
                         cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
                         seed.number = 3){
  if(!cABO %in% c('A','AB','B','O')){stop("Blood group is not valid! Valid options: 'A','AB','B','O'")}

  #require('magrittr', quietly = TRUE)

  set.seed(seed.number)

  n1 <- (100-cPRA)*100

  n.donors <- dplyr::sample_n(D10K, size = n1) |>
    dplyr::filter(bg == cABO) |>
    dplyr::mutate_if(is.numeric, as.character) |>
    dplyr::rowwise() |>
    dplyr::mutate(mmB = mmHLA(dA = c(A1,A2), dB = c(B1,B2), dDR = c(DR1,DR2),
                              cA = cA, cB = cB, cDR = cDR)['mmB'],
                  mmDR = mmHLA(dA = c(A1,A2), dB = c(B1,B2), dDR = c(DR1,DR2),
                               cA = cA, cB = cB, cDR = cDR)['mmDR'],
                  level12 = mmDR == 0 | (mmB == 0 & mmDR == 1)) |>
    dplyr::ungroup() |>
    dplyr::filter(level12) |> nrow()

  return(n.donors)

}

#' virtual PRA
#'
#' @description Computes virtual PRA (vPRA) form HLA-A, -B, -DR loci.
#' @param abs A character vector with HLA antibodies.
#' @param donors A dataframe with HLA typing for a pool of donors.
#' @return a percentual value for vPRA
#' @examples
#' vpra(abs = c('A1','A2','B5','DR4'), donors = D10K)
#' @export
#' @concept histocompatibility
vpra <- function(abs = c('A1','A2','B5','DR4'), donors = D10K){

  #require("magrittr", quietly = TRUE)

  n <- nrow(donors)

  na <- donors |>
    dplyr::mutate_at(dplyr::vars(A1,A2), function(x) paste0('A',x)) |>
    dplyr::mutate_at(dplyr::vars(B1,B2), function(x) paste0('B',x)) |>
    dplyr::mutate_at(dplyr::vars(DR1,DR2), function(x) paste0('DR',x)) |>
    dplyr::filter(A1 %in% abs | A2 %in% abs |
                    B1 %in% abs | B2 %in% abs |
                    DR1 %in% abs | DR2 %in% abs) |>
    nrow()

  res <- round(na/n * 100,1)

  return(res)

}

#' samples HLA antibodies
#'
#' @description creates a sample of HLA antibodies (abs) for a given candidate
#' according with a cPRA value.
#' @param cA candidate's HLA-A typing
#' @param cB candidate's HLA-B typing
#' @param cDR candidate's HLA-DR typing
#' @param cPRA candidate's cPRA value
#' @param origin A character value from options: 'PT', 'API', 'AFA', 'CAU' and 'HIS'
#' @param seed.number a numeric seed that will be used for random number generation.
#' @return a character vector with HLA abs.
#' @examples
#' antbs(cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
#' cPRA = 85, origin = 'PT', seed.number = 3)
#' @export
#' @concept histocompatibility
antbs <- function(cA = c('2','29'), cB = c('7','15'), cDR = c('4','7'),
                  cPRA = 85,
                  origin = 'PT', seed.number = 3){

  set.seed(seed.number)

  typing <- c(paste0('A',cA), paste0('B',cB), paste0('DR',cDR))

  if(!origin %in% c('PT','API','AFA','CAU','HIS')){stop("Origin is not valid! Valid options: 'PT','API','AFA','CAU','HIS'")}

  if(origin == 'PT') {
    valid.ags <- c(agA, agB, agDR)[!c(agA, agB, agDR) %in% typing]
    dd <- D10K
  } else {
    valid.ags <- c(agA_MNDP, agB_MNDP, agDR_MNDP)[!c(agA, agB, agDR) %in% typing]
    if(origin == 'API'){dd <- D10K_API}
    if(origin == 'AFA'){dd <- D10K_AFA}
    if(origin == 'CAU'){dd <- D10K_CAU}
    if(origin == 'HIS'){dd <- D10K_HIS}
    }

  # if(origin == 'PT'){
  #   vpra <- function(abs){vpra(abs, donors = D10K)}
  # }
  # if(origin == 'API') vpra <- function(abs) vpra(abs, donors = D10K_API)
  # if(origin == 'AFA') vpra <- function(abs) vpra(abs, donors = D10K_AFA)
  # if(origin == 'CAU') vpra <- function(abs) vpra(abs, donors = D10K_CAU)
  # if(origin == 'HIS') vpra <- function(abs) vpra(abs, donors = D10K_HIS)

  c = NULL
  if(cPRA > 0){
    for(i in 1:250){
      c[i] <- sample(valid.ags, 1, replace = F)

      if(vpra(c, donors = dd)>cPRA-3){ break }
    }
  }

  vpra <- vpra(c, donors = dd)

  list(cPRA = cPRA,
       vPRA = vpra,
       HLA = typing,
       Valid.Ags = valid.ags,
       Abs = c)

}
